waspc/waspls/src/Wasp/LSP/SignatureHelp.hs

```diff
@@ -0,0 +1,261 @@
+-- dictionary, list, tuple, or starting a new field/value in the container.
+signatureHelpTriggerCharacters :: Maybe [Char]
+signatureHelpTriggerCharacters = Just "{[(:,"
+
+-- | Configuration for 'LSP.Options', used in "Wasp.LSP.Server".
+--
+-- When a client is already displaying signature help, typing one of these
+-- characters will cause it to update the displayed signature by sending a new
+-- SignatureHelp request. It is configured so that signatures will update when
+-- ending a dictionary, list, or  tuple.
+--
+-- NOTE: 'signatureHelpTriggerCharacters' are also counted as retrigger characters.
+signatureHelpRetriggerCharacters :: Maybe [Char]
+signatureHelpRetriggerCharacters = Just "}])"
+
+-- | Get 'LSP.SignatureHelp' at a position in the wasp file. The signature
+-- displays the type of the "container" that the position is in, if any.
+--
+-- SignatureHelp is usually displayed as a popup near the text you are typing,
+-- and highlights the part of the signature you are writing.
+--
+-- A container is an expression that holds other values, such as dictionaries
+-- and lists.
+--
+-- The parameter field of the signature is used for which part of the container
+-- the position is within, such as a key for a dictionary.
+getSignatureHelpAtPosition ::
+  (MonadState ServerState m, MonadLog m) =>
+  LSP.Position ->
+  m LSP.SignatureHelp
+getSignatureHelpAtPosition position = do
+  src <- gets (^. currentWaspSource)
+  gets (^. cst) >>= \case
+    Nothing ->
+      -- No CST in the server state, can't create a signature.
+      return emptyHelp
+    Just syntax -> do
+      let offset = lspPositionToOffset src position
+      let location = toOffset offset (fromSyntaxForest syntax)
+      findSignatureAtLocation src location >>= \case
+        Nothing -> do
+          logM "[getSignatureHelpAtPosition] no signature found"
+          return emptyHelp
+        Just signature -> do
+          logM "[getSignatureHelpAtPosition] found a signature"
+          let signatureFragments = signatureToFragments signature
+          let params = getParamsFromFragments signatureFragments
+          let activeParam = signatureParam signature >>= findActiveParameterIndex signatureFragments
+          logM $ "[getSignatureHelpAtPosition] at param idx = " ++ show activeParam ++ " params = " ++ show params
+          return $
+            LSP.SignatureHelp
+              { _signatures =
+                  LSP.List
+                    [ LSP.SignatureInformation
+                        { _parameters = Just $ LSP.List params,
+                          -- NOTE: VSCode higlights the 0th parameter if
+                          -- activeParameter is Nothing, so it's set to -1 when
+                          -- we don't want any param highlighted.
+                          _activeParameter = Just $ fromMaybe (-1) activeParam,
+                          _label = Text.pack $ showFragments signatureFragments,
+                          _documentation = Nothing
```

Comment:
```haskell
LSP.SignatureHelpDocMarkup $
    LSP.MarkupContent {
        LSP._kind = LSP.MkMarkdown,
        LSP._value = Text.pack $ "```\n" ++ showFragments signatureFragments ++ "\n```"
    }
```

Reply:
<img width="615" alt="Screenshot 2023-06-06 at 11 37 07" src="https://github.com/wasp-lang/wasp/assets/2223680/47245cc0-9f50-4731-8250-3b1c303f21ab">

Excuse my poor attempt, but it could work

Reply:
This feels too much like a VSCode-specific hack to be worth it, especially since it's just duplicating information. Other editors with LSP clients seem to support multi-line signatures. Helix for example:

![image](https://github.com/wasp-lang/wasp/assets/22552467/dc4ce6a4-0419-4f6b-91a4-139b223ed7d5)

And if we look at how other LSP servers use SignatureHelp, they run into the same problem with VSCode's single line signature strings. Rust for example:

![image](https://github.com/wasp-lang/wasp/assets/22552467/dffd9998-4785-4d7d-bee2-2ab5827cd02c)
 
I think we should stick with what the specification says to use these fields for.

Anyways, I added newlines back to the signature string so that editors that support it will display it over multiple lines.

Reply:
Oho nice nice!

Reply:
Ah that sucks. But why does it look good for Typescript then, in @infomiho 's screenshot above?